nixos/modules/services/web-apps/nextcloud-backup/default.nix

```diff
@@ -0,0 +1,113 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.nextcloudBackup;
+  nextcloudCfg = config.services.nextcloud;
+  runtimeInputs =
+    [
+      pkgs.coreutils
+      pkgs.rsync
+      pkgs.jq
+      nextcloudCfg.occ
+    ]
+    ++ (
+      if nextcloudCfg.config.dbtype == "sqlite" then
+        [ pkgs.sqlite ]
+      else if nextcloudCfg.config.dbtype == "pgsql" then
+        [ config.services.postgresql.package ]
+      else if nextcloudCfg.config.dbtype == "mysql" then
+        [ config.services.mysql.package ]
+      else
+        throw "Unknown nextcloud database type ${nextcloudCfg.config.dbtype}"
+    );
+  setEnvVariables =
+    let
+      readDbpass = lib.optionalString (nextcloudCfg.config.dbpassFile != null) ''
+        "$(cat ${nextcloudCfg.config.dbpassFile})"
+      '';
+      # Mysql doesn't accept a socket path as host argument
+      dbHost =
+        if nextcloudCfg.config.dbtype == "mysql" && nextcloudCfg.database.createLocally then
+          "localhost"
+        else
+          nextcloudCfg.config.dbhost;
+      dbHostIsSocket = builtins.toString (
+        nextcloudCfg.database.createLocally
+        && (builtins.elem nextcloudCfg.config.dbtype [
+          "pgsql"
+          "mysql"
+        ])
+      );
+      backupDir = lib.optionalString (cfg.backupDir != null) "${cfg.backupDir}";
+    in
+    ''
+      export NEXTCLOUD_VERSION="${nextcloudCfg.package.version}"
+      export NEXTCLOUD_DBTYPE="${nextcloudCfg.config.dbtype}"
+      export NEXTCLOUD_DBHOST="${dbHost}"
+      export NEXTCLOUD_DBHOST_IS_SOCKET="${dbHostIsSocket}"
+      export NEXTCLOUD_DBNAME="${nextcloudCfg.config.dbname}"
+      export NEXTCLOUD_DBUSER="${nextcloudCfg.config.dbuser}"
+      export NEXTCLOUD_DBPASS="${readDbpass}"
+      export NEXTCLOUD_DATADIR="${nextcloudCfg.datadir}"
+      export NEXTCLOUD_HOME="${nextcloudCfg.home}"
+      export BACKUP_DIR="${backupDir}"
+    '';
+  backupScript = pkgs.writeShellApplication {
+    name = "nextcloud-backup";
+    inherit runtimeInputs;
+    text = ''
+      ${setEnvVariables}
+      ${lib.getExe pkgs.bash} ${./backup.sh} "$@"
+    '';
+  };
+  restoreScript = pkgs.writeShellApplication {
+    name = "nextcloud-restore";
+    runtimeInputs = runtimeInputs;
+    text = ''
+      ${setEnvVariables}
```

Comment:
You can set the env using the `runtimeEnv` argument and then set the text to the `.sh` file directly.